crates/sim2h/src/websocket/tls.rs

```diff
@@ -0,0 +1,386 @@
+extern crate native_tls;
+extern crate openssl;
+
+use crate::websocket::{FAKE_PASS, FAKE_PKCS12};
+
+use lib3h::transport::error::TransportResult;
+
+use openssl::{
+    asn1::Asn1Time,
+    bn::{BigNum, MsbOption},
+    hash::MessageDigest,
+    pkey::{PKey, Private},
+    rsa::Rsa,
+    x509::{self, X509Name, X509},
+};
+use rand::{distributions::Alphanumeric, thread_rng, Rng};
+
+// Generates a key/cert pair
+fn generate_pair() -> (PKey<Private>, x509::X509) {
+    let rsa = Rsa::generate(2048).unwrap();
+    let key = PKey::from_rsa(rsa).unwrap();
+
+    let mut name = X509Name::builder().unwrap();
+    name.append_entry_by_nid(openssl::nid::Nid::COMMONNAME, "example.com")
+        .unwrap();
+    let name = name.build();
+
+    let serial_number = {
+        let mut serial = BigNum::new().unwrap();
+        serial.rand(159, MsbOption::MAYBE_ZERO, false).unwrap();
+        serial.to_asn1_integer().unwrap()
+    };
+
+    let mut builder = X509::builder().unwrap();
+    builder.set_serial_number(&serial_number).unwrap();
+    builder.set_version(2).unwrap();
+    builder.set_subject_name(&name).unwrap();
+    builder.set_issuer_name(&name).unwrap();
+    builder.set_pubkey(&key).unwrap();
+    let not_before = Asn1Time::days_from_now(0).unwrap();
+    builder.set_not_before(&not_before).unwrap();
+    let not_after = Asn1Time::days_from_now(3650).unwrap();
+    builder.set_not_after(&not_after).unwrap();
+    builder.sign(&key, MessageDigest::sha256()).unwrap();
+
+    let cert: openssl::x509::X509 = builder.build();
+
+    (key, cert)
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
+pub struct TlsCertificate {
+    pub(in crate::websocket) pkcs12_data: Vec<u8>,
+    pub(in crate::websocket) passphrase: String,
+}
+
+impl TlsCertificate {
+    /// Creates a self-signed certificate with an entropy key and passphrase.
+    /// This makes it possible to use a TLS encrypted connection securely between two
+    /// peers using the lib3h websockt actor.
+    pub fn build_from_entropy() -> Self {
+        let (key, cert) = generate_pair();
+
+        let random_passphrase: String = thread_rng().sample_iter(&Alphanumeric).take(30).collect();
```

Comment:
it would be nice to not have remove the magic number and put it in a variable just to give a little bit of context on why it is taking 30, or even a comment would suffice

Reply:
The normal case is a user-entered passphrase... the length of the passphrase is irrelevant.

I'm not convinced we even need a real passphrase for this ephemeral case. We could set it to an empty string.

Reply:
@neonphog Is there a one line suggested change you can provide?